torch/csrc/autograd/variable.cpp

```diff
@@ -487,7 +550,144 @@ void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect
+// Tensors that do not have forward grad originally.
+void AutogradMeta::set_fw_grad(Variable& new_grad, const Variable& self, uint64_t level, bool is_inplace_op) {
+  if (!fw_grad_) {
+    // Lazy initialization
+    fw_grad_ = std::make_shared<ForwardGrad>();
+  }
+  if (fw_grad_->contains(level)) {
+    // Setting the forward grad again is only allowed if it is a no-op.
+    // We do allow this case to simplify writing codegen for inplace ops.
+    TORCH_INTERNAL_ASSERT(new_grad.defined(), "Cannot set a forward grad that is an undefined Tensor. Use "
+                          "_fw_primal(level) to get a new Tensor with this forward grad unset.");
+
+    TORCH_INTERNAL_ASSERT(is_inplace_op, "Only inplace operations can re-set the forward grad of a Tensor that "
+                          "already has one.");
+
+    TORCH_INTERNAL_ASSERT(fw_grad_->value(level).is_same(new_grad), "Cannot set a value of a forward grad if it "
+                          "already exists. Inplace operations should modify it inplace.");
+  } else {
+    // For inplace ops on a Tensor that does not already have a forward grad and is a view, we propagate
+    // the tangent to the base and ensure that the new_grad is a view of that base's tangent.
+    if (is_inplace_op && is_view_) {
+      auto this_view_meta = static_cast<DifferentiableViewMeta*>(this);
+      if (this_view_meta->has_fw_view()) {
+        auto view_info = this_view_meta->get_forward_view();
+        auto& base = view_info.base_;
+
+        if (!base.fw_grad(level).defined()) {
+          // Enforce same meta here to make sure that the view op below is always valid
+          Tensor new_base_fw_grad;
+          if (has_same_meta(new_grad, base)) {
+            // TODO extend this special case to when the underlying storage of new_grad
+            // can be re-used.
+            new_base_fw_grad = new_grad;
+          } else {
+            new_base_fw_grad = new_with_same_meta(base);
+
+            // Update new_grad to be a view of the base
+            Tensor new_fw_grad_value;
+            if (view_info.has_view_fn()) {
+              new_fw_grad_value = view_info.view_fn()(new_base_fw_grad);
+            } else {
+              new_fw_grad_value = new_base_fw_grad.as_strided(self.sizes(), self.strides(), self.storage_offset());
+            }
+
+            new_fw_grad_value.copy_(new_grad);
+            new_grad = new_fw_grad_value;
+          }
+
+          base.set_fw_grad(new_base_fw_grad, level, /* is_inplace_op */ false);
+        }
+      }
+    }
+
+    if (!has_same_meta(new_grad, self)) {
+      Tensor new_grad_with_meta = new_with_same_meta(self);
+      new_grad_with_meta.copy_(new_grad);
+      new_grad = new_grad_with_meta;
+    }
+
+    fw_grad_->set_value(new_grad, level);
+  }
+}
+
+const Variable& AutogradMeta::fw_grad(uint64_t level, const Variable& self) const {
+  bool has_no_direct_fw_grad = !(fw_grad_ && fw_grad_->value(level).defined());
```

Comment:
`fw_grad_->value(level)` takes out a lock, which means there is a race on subsqequent accesses. Save it first.

Reply:
I don't really like that a lock is being taken out here, though it's not a deal breaker for now.